torchtune/datasets/_packed.py

```diff
@@ -119,156 +117,192 @@ def _pack(self) -> None:
-            previous_sample_boundary = len(current_pack["tokens"])
-            if self.max_packs is not None and len(self.packs) >= self.max_packs:
+
+            # Keep track of previous sample boundary
+            self.previous_sample_boundary = len(current_pack["tokens"])
+
+            # If max packs is set, stop packing when we reach that number
+            if self.max_packs is not None and len(self.packs) == self.max_packs:
                 break
 
-        # Add the last pack with remaining samples that did not fit in previous
-        if len(current_pack["tokens"]) > 0 and (
-            self.max_packs is None or len(self.packs) < self.max_packs
+        # Handle the last pack if there's leftover and we haven't filled up the max packs
+        if (
+            len(current_pack["tokens"]) > 0
+            and self.max_packs is not None
+            and len(self.packs) < self.max_packs
         ):
-            current_pack = self._add_pack(
-                current_pack=current_pack, boundary=len(current_pack["tokens"])
-            )
-            assert len(current_pack["tokens"]) == 0
+            current_pack = self._add_pack(current_pack)
 
-    def _add_pack(
-        self, current_pack: Dict[str, List[int]], boundary: int
-    ) -> Dict[str, List[int]]:
-        """
-        Pad and add the current pack to self.packs and return what's remaining.
+    def _convert_to_tensors(self, pack: PACK_TYPE) -> PACK_TYPE:
+        """Converts a pack into tensors. Pack comes in as a dict of lists and is converted to tensors.
+        The only key that does not get converted is ``seq_lens``.
         """
-        # So far we've kept a list of causal masks, one for each sample in the pack.
-        # Now we need to combine them into a single mask for the entire pack.
-        packing_mask = torch.block_diag(*current_pack["mask"])
-        pack = {
-            "tokens": current_pack["tokens"][:boundary],
-            "labels": current_pack["labels"][:boundary],
-            "mask": packing_mask[:boundary, :boundary],
-            "input_pos": current_pack["input_pos"][:boundary],
-        }
-
-        # Resultant shapes after padding
-        # tokens: [max_seq_len, ]
-        # labels: [max_seq_len, ]
-        # mask: [max_seq_len, max_seq_len]
-        # input_pos: [max_seq_len, ]
-        padded_pack = self._pad_pack(pack)
-        self.packs.append(padded_pack)
-
-        # Keep sample that did not fit or got truncated for the next pack
-        updated_pack = {
-            "tokens": current_pack["tokens"][boundary:],
-            "labels": current_pack["labels"][boundary:],
-            "mask": [packing_mask[boundary:, boundary:]],
-            "input_pos": current_pack["input_pos"][boundary:],
+        return {
+            "tokens": torch.tensor(pack["tokens"]),
+            "labels": torch.tensor(pack["labels"]),
+            "input_pos": torch.tensor(pack["input_pos"]),
+            "seq_lens": pack["seq_lens"],
```

Comment:
should we just standardize on torch.tensor here, or would that make the enumerate in getitem more messy?

Reply:
Ehh, it just takes up more memory and isn't necessary.

Reply:
I am inclined to agree with @RdoubleA here. Is the memory increase just tensor metadata? If so should be pretty small, right? Advantage is that we can then scrap this whole `PACK_TYPE` business. Either way not a blocker